File: src/main/java/ch/zxseitz/j3de/math/Vector2.java
package ch.zxseitz.j3de.math;

public class Vector2 {
    public static final Vector2 ZERO = new Vector2(0f, 0f);
    public static final Vector2 ONE = new Vector2(1f, 1f);
    public static final Vector2 X = new Vector2(1f, 0f);
    public static final Vector2 Y = new Vector2(0f, 1f);

    public final float x, y;

    public Vector2(float x, float y) {
        this.x = x;
        this.y = y;
    }

    public Vector2(Vector2 a) {
        this.x = a.x;
        this.y = a.y;
    }

    public Vector2 add(Vector2 vec) {
        return new Vector2(x + vec.x, y + vec.y);
    }

    public Vector2 subtract(Vector2 vec) {
        return new Vector2(x - vec.x, y - vec.y);
    }

    public Vector2 multiply(Vector2 vec) {
        return new Vector2(x * vec.x, y * vec.y);
    }

    public Vector2 divide(Vector2 vec) {
        return new Vector2(x / vec.x, y / vec.y);
    }

    public Vector2 scale(float s) {
        return new Vector2(x * s, y * s);
    }

    public float norm() {
        return (float) Math.sqrt(normSquared());
    }

    public float normSquared() {
        return x * x + y * y;
    }

    public float dot(Vector2 v) {
        return x * v.x + y * v.y;
    }

    @Override
    public boolean equals(Object obj) {
//        if (obj instanceof Vector2 vec) {
        if (obj instanceof Vector2) {
            var vec = (Vector2) obj;
            return MathUtils.isFloatEquals(x, vec.x) && MathUtils.isFloatEquals(y, vec.y);
        }
        return false;
    }

    @Override
    public String toString() {
        return String.format("(%.3f, %.3f)", x, y);
    }
}

File: src/main/java/ch/zxseitz/j3de/math/Vector2.kt
package ch.zxseitz.j3de.math

import kotlin.math.sqrt

data class Vector2(val x : Float = 0f, val y: Float = 0f) {
    operator fun plus(v: Vector2): Vector2 {
        return Vector2(x + v.x, y + v.y)
    }
    operator fun minus(v: Vector2): Vector2 {
        return Vector2(x - v.x, y - v.y)
    }
    operator fun times(s: Float): Vector2 {
        return Vector2(x * s, y * s)
    }
    fun normSquared(): Float {
        return x * x + y * y
    }
    fun norm(): Float {
        return sqrt(normSquared())
    }
    fun normalize(): Vector2 {
        return this * (1f / norm())
    }

    override fun toString(): String {
        return "(%.3f, %.3f)".format(x, y)
    }
}
operator fun Vector2.unaryMinus() = Vector2(-x, -y)

infix fun Vector2._times(v: Vector2): Vector2 {
    return Vector2(x * v.x, y * v.y)
}
infix fun Vector2.dot(v: Vector2): Float {
    return x * v.x + y * v.y
}
